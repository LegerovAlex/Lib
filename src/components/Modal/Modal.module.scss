.backdrop {
  position: fixed;
  inset: 0;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 999;
  display: flex;
  align-items: center;
  justify-content: center;
}

.modal {
  background-color: #fff;
  border-radius: 12px;
  padding: 24px;
  max-width: 500px;
  width: 100%;
  min-height: 30vh;
  box-shadow: 0 8px 30px rgba(0, 0, 0, 0.15);
  position: relative;
}

.closeButton {
  position: absolute;
  top: 5px;
  right: 5px;
  background: transparent;
  border: none;
  font-size: 18px;
  cursor: pointer;
  color: #1976d2;

  &:hover {
    color: #000;
  }
}
