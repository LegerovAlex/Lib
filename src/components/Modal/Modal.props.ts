import { ReactNode } from 'react';

export interface ModalProps {
  IsOpen: boolean;
  onClose: () => void;
  children: ReactNode;
}
