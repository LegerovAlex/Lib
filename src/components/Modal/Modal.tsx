import { FC, useEffect } from 'react';
import { ModalProps } from './Modal.props';
import { createPortal } from 'react-dom';
import styles from './Modal.module.scss';

export const Modal: FC<ModalProps> = ({ onClose, isOpen, children }) => {
  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        onClose();
      }
    };

    if (isOpen) {
      document.addEventListener('keydown', handleKeyDown);
      document.body.style.overflow = 'hidden';
    }

    return () => {
      document.removeEventListener('keydown', handleKeyDown);
      document.body.style.overflow = '';
    };
  }, [isOpen, onClose]);

  if (!isOpen) return null;

  const handleBackdropClick = (event: React.MouseEvent<HTMLDivElement>) => {
    if (event.currentTarget === event.target) {
      onClose();
    }
  };

  return createPortal(
    <div
      role="dialog"
      onClick={handleBackdropClick}
      className={styles.backdrop}
      data-testid="backdrop"
    >
      <div className={styles.modal}>
        <button onClick={onClose} className={styles.closeButton}>
          Ñ…
        </button>
        {children}
      </div>
    </div>,
    document.getElementById('modal-root'),
  );
};
