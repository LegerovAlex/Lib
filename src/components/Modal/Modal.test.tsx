import { fireEvent, render, screen } from '@testing-library/react';
import { ModalProps } from './Modal.props';
import { Modal } from './Modal';
import userEvent from '@testing-library/user-event';

export interface RenderModalResult {
  modal: HTMLElement | null;
  backdrop: HTMLElement | null;
  container: HTMLElement | null;
  closeButton: HTMLElement | null;
  onClose: jest.Mock;
}

export function renderModal(props: Partial<ModalProps> = {}): RenderModalResult {
  const onClose = jest.fn();

  const container = document.createElement('div');
  container.setAttribute('id', 'modal-root');
  document.body.appendChild(container);

  render(
    <Modal onClose={onClose} isOpen={props.isOpen}>
      {props.children ?? 'Modal Content'}
    </Modal>,
  );

  const modal = screen.queryByRole('dialog');
  const backdrop = screen.queryByTestId('backdrop');
  const closeButton = screen.queryByRole('button', { name: /Ñ…/i });

  return { modal, backdrop, closeButton, onClose, container };
}

describe('Modal', () => {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  it('renders modal when isOpen is true', () => {
    const { modal, backdrop, closeButton } = renderModal({ isOpen: true });
    expect(modal).toBeInTheDocument();
    expect(backdrop).toBeInTheDocument();
    expect(closeButton).toBeInTheDocument();
    expect(modal).toHaveTextContent('Modal Content');
  });
  it('does not render modal when isOpen is false', () => {
    const { modal, backdrop, closeButton } = renderModal({ isOpen: false });
    expect(modal).not.toBeInTheDocument();
    expect(backdrop).not.toBeInTheDocument();
    expect(closeButton).not.toBeInTheDocument();
  });
  it('renders custom children', () => {
    const { modal } = renderModal({ isOpen: true, children: 'Custom Content' });
    expect(modal).toBeInTheDocument();
    expect(modal).toHaveTextContent('Custom Content');
  });
  it('renders in portal container', () => {
    const { container, modal } = renderModal({ isOpen: true });
    expect(container).toContainElement(modal);
  });
  it('calls onClose when Escape key is pressed', () => {
    const { onClose } = renderModal({ isOpen: true });
    fireEvent.keyDown(document, { key: 'Escape', code: 'Escape' });
    expect(onClose).toHaveBeenCalledTimes(1);
  });
  it('calls onClose when backdrop is clicked', async () => {
    const { backdrop, onClose } = renderModal({ isOpen: true });
    await userEvent.click(backdrop);
    expect(onClose).toHaveBeenCalledTimes(1);
  });
  it('does not call onClose when clicking inside modal content', async () => {
    const { onClose } = renderModal({ isOpen: true });
    const modalContent = screen.getByText(/modal content/i);
    await userEvent.click(modalContent);
    expect(onClose).not.toHaveBeenCalled();
  });
  it('calls onClose when close button is clicked', async () => {
    const { closeButton, onClose } = renderModal({ isOpen: true });
    await userEvent.click(closeButton);
    expect(onClose).toHaveBeenCalledTimes(1);
  });
  it('matches snapshot when open', () => {
    const { container } = renderModal({ isOpen: true });
    expect(container).toMatchSnapshot();
  });
  it('matches snapshot when closed', () => {
    const { container } = renderModal({ isOpen: false });
    expect(container).toMatchSnapshot();
  });
});
