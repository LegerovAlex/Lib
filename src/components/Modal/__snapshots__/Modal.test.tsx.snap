// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Modal matches snapshot when closed 1`] = `
<div
  id="modal-root"
/>
`;

exports[`Modal matches snapshot when open 1`] = `
<div
  id="modal-root"
>
  <div
    class="backdrop"
    data-testid="backdrop"
    role="dialog"
  >
    <div
      class="modal"
    >
      <button
        class="closeButton"
      >
        Ñ…
      </button>
      Modal Content
    </div>
  </div>
</div>
`;
